byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs

```diff
@@ -29,6 +34,44 @@ moneyIsConstant = withTests 200 . property $ do
   (st0, st) <- firstAndLastState <$> forAll (trace @UTXOW 500)
   reserves st0 + balance (utxo st0) === reserves st + balance (utxo st)
 
+-- | Check that a UTxO has no common inputs with a transaction
+noCommonInputs :: UTxO -> Tx -> Bool
+noCommonInputs utxo0 tx = dom (txouts tx) ∩ dom utxo0 == empty
+
+-- | Check that there is no double spending
+noDoubleSpending :: Property
+noDoubleSpending = property $ do
+  t <- forAll (trace @UTXO 500)
+  let
+    utxo0 = fst . firstAndLastState $ t :: State UTXO
+    txs = traceSignals OldestFirst t :: [Signal UTXO]
+  when (all (noCommonInputs $ utxo utxo0) txs) $ do
+    (all (noCommonInputsTxs txs) (zip txs [0 .. ])) === True
+   where
+    noCommonInputsTxs :: [Tx] -> (Tx, Int) -> Bool
+    noCommonInputsTxs txs (tx, i) =
+      let ins = txins' tx
+      in all (\txj -> txins' txj ∩ ins == empty) (take i txs)
+
+    txins' :: Tx -> Set TxIn
+    txins' = fromList . txins
+
+-- | Check that UTxO is outputs minus inputs
+utxoDiff :: Property
+utxoDiff = property $ do
+  t <- forAll (trace @UTXO 500)
+  let
+    (utxo0, utxoSt) = (\(f, s) -> (utxo f, utxo s)) . firstAndLastState $ t :: (UTxO, UTxO)
```

Comment:
You could also:

```haskell
import           Control.Arrow ((***))

    (utxo0, utxoSt) = (utxo *** utxo) . firstAndLastState $ t
```

or

```haskell
(UTxOState {utxo = utxo0}, UTxOState {utxo = utxoSt}) = firstAndLastState t
```

whatever makes sense to you.

BTW, we don't need the type annotation it seems.

Reply:
Thanks for the arrow hint! I added it in the upcoming commit.

When it comes to the type annotation, I know it type-checks without them; they're here to ease understanding the code (e.g., to myself in a month from now).

Reply:
> When it comes to the type annotation, I know it type-checks without them; they're here to ease understanding the code (e.g., to myself in a month from now).

:+1: